app/plugin_api.go

```diff
@@ -45,6 +45,12 @@ func (api *PluginAPI) LoadPluginConfiguration(dest interface{}) error {
 		finalConfig[strings.ToLower(setting)] = value
 	}
 
+	if api.ValidateConfig(finalConfig) {
```

Comment:
1/5 Should the check be better placed in `Environment) Activate()` https://github.com/mattermost/mattermost-server/blob/9bf5687311093f9773f42da3848fc03d8abe5d1b/plugin/environment.go#L131

Seems that the desirable outcome is that 
(a) An attempt to install a plugin into an invalid configuration should fail (we may even manage to show the right error message!).
(b) Plugin "configuration required" errors should be clearly displayed on startup.

I can see how placing the check here implements (2), not sure if it covers (1) - by the time this code is executed, the plugin is already installed "successfully"?

Reply:
Moving to `func (env *Environment) Activate(id string) (manifest *model.Manifest, activated bool, reterr error)` makes sense if you want to prevent startup on invalid configuration.
(a) admin friendly error message like I suggested to @jasonblais: `2\. Error: for plugin [com.mattermost.sample-plugin] required key [someKey] is missing, please add configuration for key [someKey] within the following JSON path [Pluginsettings->Plugins->com.mattermost.sample-plugin->someKey] of type [bool] associated help text is [When true, an extra feature will be enabled!] within [config.json]`
(b) this would be fixed by moving to Activate right ?

Reply:
Moving the `func (m *Manifest) ValidateConfig(config map[string]interface{}) error` to `func (env *Environment) Activate(id string, config *model.Config) (manifest *model.Manifest, activated bool, reterr error)` has two side effects, `ValidateConfig` is now part of the `Manifest` interface and `Activate` now requires `*model.config` as a parameter in order that `ValidateConfig` may obtain the plugin configuration via the server config